File: kindlife_app/kindlife_app/report/current_item_price/current_item_price.py
import frappe
from frappe import _

def execute(filters=None):
    if not filters:
        filters = {}
    
    # Validate mandatory filter
    if not filters.get('price_list'):
        frappe.throw(_('Price List is mandatory'))
    
    columns = get_columns(filters)
    data = get_data(filters)
    
    return columns, data

def get_columns(filters):
    # Get the currency from the selected price list
    
    return [
        {
            "label": _("Inventory ID"),
            "fieldname": "item_code",
            "fieldtype": "Link",
            "options": "Item",
            "width": 120
        },
        {
            "label": _("UOM"),
            "fieldname": "uom",
            "fieldtype": "Link",
            "options": "UOM",
            "width": 80
        },
        {
            "label": _("Item Name"),
            "fieldname": "item_name",
            "fieldtype": "Data",
            "width": 200
        },
        {
            "label": _("MRP"),
            "fieldname": "custom_buying_price_formatted",
            "fieldtype": "Data",
            "width": 100
        },
        {
            "label": _("Discount On"),
            "fieldname": "custom_discount_on",
            "fieldtype": "Data",
            "width": 100
        },
        {
            "label": _("Discount %"),
            "fieldname": "custom_discount_percentage",
            "fieldtype": "Data",
            "width": 80
        },
        {
            "label": _("Discount Amount"),
            "fieldname": "custom_discount_amount",
            "fieldtype": "Data",
            "width": 80
        },
        {
            "label": _("Buying Price"),
            "fieldname": "price_list_rate_formatted",
            "fieldtype": "Data",
            "width": 100
        },
        {
            "label": _("Approved By"),
            "fieldname": "custom_approved_by",
            "fieldtype": "Link",
            "options": "User",
            "width": 120
        },
        {
            "label": _("Approved Date"),
            "fieldname": "custom_approved_at",
            "fieldtype": "Date",
            "width": 130
        },
        {
            "label": _("Valid From"),
            "fieldname": "valid_from",
            "fieldtype": "Date",
            "width": 130
        },
        # {
        #     "label": _("Pending Approvals"),
        #     "fieldname": "pending_count",
        #     "fieldtype": "Int",
        #     "width": 120
        # },
        {
            "label": _("Actions"),
            "fieldname": "actions",
            "fieldtype": "Data",
            "width": 150
        }
    ]

def get_data(filters):
    price_list = filters.get('price_list')
        
    # Get current valid item prices
    query = """
        SELECT 
			ip.item_code,
			i.stock_uom as uom,
			i.item_name,
			ip.price_list_rate,
			ip.custom_buying_price,
            ip.custom_discount_on,
            ip.custom_discount_percentage,
            ip.custom_discount_amount,
			ip.custom_approved_by,
			ip.custom_approved_at,
			ip.valid_from,
			ip.name as item_price_name,
			ip.currency,
			CURDATE() as curdate
		FROM 
			`tabItem Price` ip
		LEFT JOIN 
			`tabItem` i ON ip.item_code = i.name
		
		WHERE 
			ip.price_list = %(price_list)s
            AND ip.custom_disable = 0
			AND ip.workflow_state='Approved'
			AND ip.valid_from = (
				SELECT MAX(ip2.valid_from)
				FROM `tabItem Price` ip2
				WHERE ip2.item_code = ip.item_code
				AND ip2.price_list = ip.price_list
				AND ip2.valid_from <= CURDATE()
				AND ip2.workflow_state='Approved'
                AND ip2.custom_disable = 0
			)
			AND ip.name = (
				SELECT ip3.name
				FROM `tabItem Price` ip3
				WHERE ip3.item_code = ip.item_code
				AND ip3.price_list = ip.price_list
				AND ip3.workflow_state = 'Approved'
				AND ip3.valid_from = ip.valid_from
                AND ip3.custom_disable = 0
				ORDER BY ip3.modified DESC
				LIMIT 1
			)
		ORDER BY 
			ip.item_code
    """
    
    data = frappe.db.sql(query, {'price_list': price_list}, as_dict=1)
    print("data",data)
    
    # Add pending approval count for each item
    for row in data:
        # Format currency values with proper currency symbol
        if row.get('currency'):
            currency_symbol = frappe.db.get_value('Currency', row.get('currency'), 'symbol') or row.get('currency')
            
            # Format the currency values
            if row.get('custom_buying_price'):
                row['custom_buying_price_formatted'] = f"{frappe.utils.fmt_money(row['custom_buying_price'], currency=row.get('currency'))}"
            else:
                row['custom_buying_price_formatted'] = f"{currency_symbol} 0.00"
                
            if row.get('price_list_rate'):
                row['price_list_rate_formatted'] = f"{frappe.utils.fmt_money(row['price_list_rate'], currency=row.get('currency'))}"
            else:
                row['price_list_rate_formatted'] = f"{currency_symbol} 0.00"
        else:
            # Fallback if currency is not available
            row['custom_buying_price_formatted'] = str(row.get('custom_buying_price', 0))
            row['price_list_rate_formatted'] = str(row.get('price_list_rate', 0))
        
        pending_count = frappe.db.count('Item Price', {
            'item_code': row.item_code,
            'price_list': price_list,
            'docstatus': 0,
            'workflow_state': 'Pending'
        })
        row['pending_count'] = pending_count
        
        # Add action buttons HTML
        # Add combined actions HTML with pending count and change price button
        pending_badge = ""
        if row['pending_count'] > 0:
            pending_badge = f"""
                <span class="badge badge-warning" 
                     style="cursor: pointer; font-size: 10px; padding: 3px 6px; border-radius: 10px; margin-right: 5px;"
                     onclick="viewPendingPrices('{row.item_code}', '{price_list}')"
                     title="Click to view {row['pending_count']} pending approval(s)">
                    <i class="fa fa-clock-o" style="font-size: 9px;"></i> {row['pending_count']}
                </span>
            """
        else:
            pending_badge = f"""
                <span class="badge badge-success" 
                      style="font-size: 10px; padding: 3px 6px; border-radius: 10px; margin-right: 5px;">
                    <i class="fa fa-check" style="font-size: 9px;"></i> 0
                </span>
            """
        
        row['actions'] = f"""
            <div style="display: flex; align-items: center; justify-content: center; gap: 5px;">
                {pending_badge}
                <button class='btn btn-xs btn-primary' 
                        style='padding: 2px 6px; font-size: 10px; min-width: auto;'
                        onclick='changePrice("{row.item_code}", "{price_list}")'
                        title='Change Price'>
                    <i class="fa fa-edit"></i>
                </button>
            </div>
        """
    
    return data

File: kindlife_app/kindlife_app/report/current_item_price/current_item_price.js
frappe.query_reports["Current Item Price"] = {
    "filters": [
        {
            "fieldname": "price_list",
            "label": __("Price List"),
            "fieldtype": "Link",
            "options": "Price List",
            "reqd": 1
        }
    ],

    "onload": function (report) {
        // Add button for pending approval report
        report.page.add_inner_button(__('Pending Approval Prices'), function () {
            frappe.set_route('query-report', 'Pending Item Price Approval', {
                'price_list': frappe.query_report.get_filter_value('price_list')
            });
        });
    },



    //    "formatter": function(value, row, column, data, default_formatter) {
    //         if (column.fieldname == "pending_count") {
    //             if (value > 0) {
    //                 return `<div style="text-align: center;">
    //                             <span class="badge badge-warning" 
    //                                  style="cursor: pointer; font-size: 10px; padding: 3px 6px; border-radius: 10px;"
    //                                  onclick="viewPendingPrices('${data.item_code}', '${frappe.query_report.get_filter_value('price_list')}')"
    //                                  title="Click to view ${value} pending approval(s)">
    //                                 <i class="fa fa-clock-o" style="font-size: 9px;"></i> ${value}
    //                             </span>
    //                         </div>`;
    //             } else {
    //                 return `<div style="text-align: center;">
    //                             <span class="badge badge-success" 
    //                                   style="font-size: 10px; padding: 3px 6px; border-radius: 10px;">
    //                                 <i class="fa fa-check" style="font-size: 9px;"></i> 0
    //                             </span>
    //                         </div>`;
    //             }
    //         }
    //         return default_formatter(value, row, column, data);
    //     }
};

// Global functions for actions
window.changePrice = function (item_code, price_list) {
    let dialog = new frappe.ui.Dialog({
        title: __('Change Item Price'),
        fields: [
            {
                fieldname: 'item_code',
                label: __('Item Code'),
                fieldtype: 'Link',
                options: 'Item',
                default: item_code,
                read_only: 1
            },
            {
                fieldname: 'price_list',
                label: __('Price List'),
                fieldtype: 'Link',
                options: 'Price List',
                default: price_list,
                read_only: 1
            },
            {
                fieldname: 'new_buying_price',
                label: __('New Buying Price'),
                fieldtype: 'Currency',
                reqd: 1,
                onchange: function () {
                    calculate_discount_dialog(this.dialog);
                }
            },
            {
                fieldname: 'new_mrp',
                label: __('New MRP'),
                fieldtype: 'Currency',
                reqd: 1,
                onchange: function () {
                    calculate_discount_dialog(this.dialog);
                }
            },
            {
                fieldname: 'custom_discount_on',
                label: __('Discount On'),
                fieldtype: 'Select',
                options: 'Selling Price\nMRP',
                default: 'Selling Price',
                onchange: function () {
                    calculate_discount_dialog(this.dialog);
                }
            },
            {
                fieldname: 'custom_discount_percentage',
                label: __('Discount Percentage'),
                fieldtype: 'Float',
                onchange: function () {
                    calculate_discount_dialog(this.dialog);
                }
            },
            {
                fieldname: 'custom_discount_amount',
                label: __('Discount Amount'),
                fieldtype: 'Float',
                onchange: function () {
                    // reverse_calculate_discount_dialog(this.dialog); // Optional: Implement if needed
                }
            },
            {
                fieldname: 'valid_from',
                label: __('Valid From'),
                fieldtype: 'Date',
                default: frappe.datetime.get_today(),
                reqd: 1
            }
        ],
        primary_action_label: __('Save and Submit'),
        primary_action: function (values) {
            frappe.call({
                method: 'kindlife_app.api.item_price.create_item_price_for_approval',
                args: values,
                callback: function (r) {
                    if (r.message) {
                        frappe.msgprint(__('Price changes submitted'));
                        dialog.hide();
                        frappe.query_report.refresh();
                    }
                }
            });
        }
    });

    dialog.show();
};

function calculate_discount_dialog(dialog) {
    let values = dialog.get_values();
    let custom_discount_on = values.custom_discount_on;
    let discount_percentage = flt(values.custom_discount_percentage);

    if (!discount_percentage) return;

    if (custom_discount_on === 'MRP') {
        let custom_buying_price = flt(values.new_mrp);

        if (custom_buying_price > 0) {
            let discount_amount = flt(custom_buying_price * (discount_percentage / 100.0), 2);
            dialog.set_value('custom_discount_amount', discount_amount);

            // Update buying price (price_list_rate)
            let new_rate = flt(custom_buying_price - discount_amount, 2);
            dialog.set_value('new_buying_price', new_rate);
        }
    } else if (custom_discount_on === 'Selling Price') {
        let price_list_rate = flt(values.new_buying_price);

        if (price_list_rate > 0) {
            let discount_amount = flt(price_list_rate * (discount_percentage / 100.0), 2);
            dialog.set_value('custom_discount_amount', discount_amount);
            // Do NOT update price_list_rate recursively
        }
    }
}

window.viewPendingPrices = function (item_code, price_list) {
    frappe.set_route('query-report', 'Pending Item Price Approval', {
        'price_list': price_list,
        'item_code': item_code,
    });
};